### PR TITLE
[GSOC] : Plugin Architecture for Admin 

### LINKED ISSUE #356: Feature Request : [Features as Plugins] Implement UI for AddOnStore with Proper Server Calls
Details will be added soon
**Is your feature request related to a problem? Please describe.**

[Latent features in the mobile app should be enabled by installing plugins via the Admin web portal. The API would need to detect the existence of a plugin, and automatically query the plugin for authorization to do some action. When authorized, the mobile app will display new capabilities. This project has two parts:
](https://palisadoesfoundation.github.io/talawa-docs/docs/internships/internship-ideas/#admin-web-portal-plugin---newsfeed-advertising)
    

**Describe the solution you'd like**

- [x] The design and development of the plugin architecture

  

**Describe alternatives you've considered**

Currently, I don't have decided on any other alternatives but alternatives can be suggested by mentors

**Approach to be followed (optional)**
tasks  :

- [x] Creation of Store UI with proper API calls
- [x] Creation of Plugin Cards UI with proper API calls
- [x] Creation of necessary queries in `mutations.ts` and `queries.ts` files 

**Additional context**
The screenshots and working video is provided in the pull request description #355

### PR DESCRIPTION
**Pull request by SiddheshKukade**
**What kind of change does this PR introduce?**

Feature Request : #356 

**Issue Number:**

Fixes #356 

**Did you add tests for your changes?**

Yes added some tests. but will add more detailed ones after the changes are finalized

**Snapshots/Videos:**

This video shows how admin of an org will enable , disable  and search plugin from store plugin UI

https://user-images.githubusercontent.com/65951872/179347117-329d077e-e490-4060-af22-17a9824194c9.mp4

#### If Video is Not visible use the YouTube Link : https://www.youtube.com/watch?v=o4XCwB_9lm0
**If relevant, did you update the documentation?**

Not Needed.

**Summary**

- Refractored Store UI shown in video abve
- Added new Queries for Plugins
```js
/**
 * @name PLUGIN_GET
 * @description used to fetch list of plugins
 */
export const PLUGIN_GET = gql`
  query getPluginList {
    getPlugins {
      _id
      pluginName
      pluginCreatedBy
      pluginDesc
      pluginInstallStatus
    }
  }
`;
```

- Added Mutations to add new plugins 
```js
/**
 * @name ADD_PLUGIN_MUTATION
 * @description used  `createPlugin` to add new Plugin in database
 */
export const ADD_PLUGIN_MUTATION = gql`
  mutation add_plugin_mutation(
    $pluginName: String!
    $pluginCreatedBy: String!
    $pluginDesc: String!
    $pluginInstallStatus: Boolean!
    $installedOrgs: [ID!]
  ) {
    createPlugin(
      pluginName: $pluginName
      pluginCreatedBy: $pluginCreatedBy
      pluginDesc: $pluginDesc
      pluginInstallStatus: $pluginInstallStatus
      installedOrgs: $installedOrgs
    ) {
      _id
      pluginName
      pluginCreatedBy
      pluginDesc
      pluginInstallStatus
      installedOrgs
    }
  }
`;
```

- Added Mutations to modify plugin properties 
```js
/**
 * @name UPDATE_INSTALL_STATUS_PLUGIN_MUTATION
 * @description used to toggle `installStatus` (boolean value) of a Plugin
 */
export const UPDATE_INSTALL_STATUS_PLUGIN_MUTATION = gql`
  mutation update_install_status_plugin_mutation($id: ID!, $status: Boolean!) {
    updateTempPluginStatus(id: $id, status: $status) {
      _id
      pluginName
      pluginCreatedBy
      pluginDesc
      pluginInstallStatus
    }
  }
`;
/**
 * @name UPDATE_ORG_STATUS_PLUGIN_MUTATION
 * @description used  `updateTempPluginInstalledOrgs`to add or remove the current Organization the in the plugin list `installedOrgs`
 */
export const UPDATE_ORG_STATUS_PLUGIN_MUTATION = gql`
  mutation update_install_status_plugin_mutation($id: ID!, $orgId: ID!) {
    updateTempPluginInstalledOrgs(id: $id, orgId: $orgId) {
      _id
      pluginName
      pluginCreatedBy
      pluginDesc
      pluginInstallStatus
      installedOrgs
    }
  }
`;
```

**Does this PR introduce a breaking change?**

No. Just some refractorings on the current store UI and some additions as well.

**Other information**

Refactored Existing code added  some new API GraphQL queries and mutations
**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

File: src/App.css [hunk 1]
```diff
@@ -1,47 +1,47 @@
...
-    position: relative;
-    text-indent: -9999em;
-    border-top: 1.1em solid rgba(255, 255, 255, 0.2);
-    border-right: 1.1em solid rgba(255, 255, 255, 0.2);
-    border-bottom: 1.1em solid rgba(255, 255, 255, 0.2);
-    border-left: 1.1em solid #febc59;
-    -webkit-transform: translateZ(0);
-    -ms-transform: translateZ(0);
-    transform: translateZ(0);
-    -webkit-animation: load8 1.1s infinite linear;
```
Comment (SiddheshKukade) [inline src/App.css:21]: These changes are done by  code formatter

File: src/components/AddOn/core/AddOnEntry/AddOnEntry.test.tsx [hunk 2]
```diff
@@ -5,25 +5,27 @@ import { BrowserRouter } from 'react-router-dom';
...
+  //   createdBy: 'string',
+  //   component: 'string',
+  //   installed: true,
+  //   configurable: true,
+  //   modified: true,
+  //   isInstalled: true,
+  //   getInstalledPlugins: () => {
+  //     return { sample: 'sample' };
+  //   },
+  // };
```
Comment (SiddheshKukade) [inline src/components/AddOn/core/AddOnEntry/AddOnEntry.test.tsx:23]: comments are added to avoid failing tests because the structure of plugins is changed

**Comment by SiddheshKukade** [comment]
Dear Mentors, @DMills27 @xsh2047 
I would like to you have your thoughts on current changes please suggest if anything needs to be changed or refractored

**Comment by palisadoes** [comment]
@SiddheshKukade The code coverage has fallen significantly. It seems as if it's caused by your proposed changes. Please resolve this as part of your PR.

image

**Comment by SiddheshKukade** [comment]
@palisadoes sir I have added some tests to increase the coverage. Will add more cases to increase the coverage ones the changes are finalized.

**Review by DMills27 — APPROVED** [review]
Good job on providing a detailed description and video make it easier for the reviewers to test themselves. It is fine as-is, however, I suggest you add more tests to increase the coverage further at a future date.

**Comment by SiddheshKukade** [comment]
> 

@DMills27 Thanks for the compliments &  surely I will add test in current week👍🏻

**Comment by SiddheshKukade** [comment]
@JamaicanFriedChicken @xsh2047  please review so that I can move further in the project. Also please suggest for changes or any betterment if needed 👍🏻

**Comment by palisadoes** [comment]
@DMills27 @JamaicanFriedChicken Please review. This has been in the queue for over 3 weeks

**Review by yasharth291 — APPROVED** [review]

**Pull request merged**